### ISSUE TITLE
ec2_asg module sporadically fails to get async notification

### ISSUE THREAD

**Issue by srkiNZ84** [ansible/ansible-modules-core#3625]
##### ISSUE TYPE
- Bug Report
##### COMPONENT NAME

ec2_asg
##### ANSIBLE VERSION

```
ansible 2.0.0.2
```
##### OS / ENVIRONMENT

Running from: Ubuntu Linux
Managing: Ubuntu Linux
##### SUMMARY

We (seeming randomly) get cases where we'll do a blue/green deploy using the "ec2_asg" module and have an async task waiting for the module to return a result. The task that waits for the result never gets the async notification and therefore fails despite the deploy succeeding.
##### STEPS TO REPRODUCE
- Create a new launch config (our new "blue" deploy)
- Run the "ec2_asg" task with the new launch config (with async set and "poll: 0")
- Have a task later in the playbook waiting on the result
- Confirm that the deploy succeeds in AWS (new instances brought up, old ones terminated)
- See that the "async_status" job never gets the notification that the deploy has happened

```
- name: Create integration tier launch configuration
  ec2_lc:
    name: "{{ environ }}-int-launch-config-{{ current_time }}"
    [OMITTED FOR BREVITY]
  register: int_launch_configuration

- name: Create Integration Autoscaling group
  ec2_asg:
    name: "{{ environ }}-int-asg"
    launch_config_name: "{{ environ }}-int-launch-config-{{ current_time }}"
    vpc_zone_identifier: "{{ int_subnets }}"
    health_check_type: "ELB"
    health_check_period: 400
    termination_policies: "OldestInstance"
    replace_all_instances: yes
    wait_timeout: 2400
    replace_batch_size: "{{ int_replace_batch_size }}"
  async: 1000
  poll: 0
  register: int_asg_sleeper

  - name: 'int ASG - check on fire and forget task'
    async_status: jid={{ int_asg_sleeper.ansible_job_id }}
    register: int_asg_job_result
    until: int_asg_job_result.finished
    retries: 60
    delay: 15
```
##### EXPECTED RESULTS

Expected that when the deploy succeeds and the "old" instances are terminated, the Async job gets the message and reports success.
##### ACTUAL RESULTS

It appears that the "file" mechanism which Python/Ansible use for checking on the status of background jobs fails and the file is never populated, despite the job finishing. Therefore the job polling the file times out eventually.

```
08:03:34.063 TASK [launch-config : int ASG - check on fire and forget task] *****************
08:03:34.130 fatal: [localhost]: FAILED! => {"failed": true, "msg": "ERROR! The conditional check 'int_asg_job_result.finished' failed. The error was: ERROR! error while evaluating conditional (int_asg_job_result.finished): ERROR! 'dict object' has no attribute 'finished'"}
```

**Comment by srkiNZ84** [comment]
I should also mention that previously we didn't do this using the "async" mechanism, rather we did it synchronously and didn't see this issue.
The problem started happening when we implemented async in order to deploy to three auto scaling groups in parallel.

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/jctanner/ansibullbot/blob/ISSUE_TRIAGE_CLASS/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by electrofelix** [comment]
Notice that your task is not timing out, instead it's hitting an exception about the `finished` field not existing. This is because initially that field is not defined until the task is complete.

I suggest changing the until to use the following format and retest:
```
  until: int_asg_job_result.finished is defined and int_asg_job_result.finished
```

The need to do this has possibly has been fixed by https://github.com/ansible/ansible-modules-core/commit/b75f305079c89ba44358890db7145dd507b3dc92, though it could still occur if the task hit an exception since the json written out in those cases do not contain the 'finished' field.

I believe the problem stems from the initial json written out would have been:
```
{ "started" : 1, "ansible_job_id" : jid, "results_file" : job_path }
```
followed by:
```
{ "started" : 1, "ansible_job_id" : jid }
```

Notice there is no finish field there, so the until above would not see 'finished' if it checked too soon.

**Comment by electrofelix** [comment]
Think this might be a duplicate of #2374

**Comment by ansibot** [comment]
This repository has been locked. All new issues and pull requests should be filed in https://github.com/ansible/ansible

Please read through the <a href="http://docs.ansible.com/ansible/dev_guide/repomerge.html#move-issues-and-prs-to-new-repo">repomerge</a> page in the dev guide. The guide contains links to tools which automatically move your issue or pull request to the ansible/ansible repo.

**Comment by ansibot** [comment]
This issue was migrated to https://github.com/ansible/ansible/issues/29801
